akka-http-core/src/main/scala/akka/http/impl/engine/client/OutgoingConnectionBlueprint.scala

```diff
@@ -149,6 +133,68 @@ private[http] object OutgoingConnectionBlueprint {
 
   import ParserOutput._
 
+  private final class ResponsePrep(parserSettings: ParserSettings, eagerClose: Boolean)
+    extends GraphStage[FlowShape[ResponseOutput, HttpResponse]] {
+
+    private val in = Inlet[ResponseOutput]("ResponsePrep.in")
+    private val out = Outlet[HttpResponse]("ResponsePrep.out")
+
+    val shape = new FlowShape(in, out)
+
+    override def createLogic(effectiveAttributes: Attributes) = new GraphStageLogic(shape) {
+      private val idle = new InHandler {
+        def onPush(): Unit = grab(in) match {
+          case ResponseStart(statusCode, protocol, headers, entityCreator, _) ⇒
+            val entity = createEntity(entityCreator) withSizeLimit parserSettings.maxContentLength
+            push(out, HttpResponse(statusCode, headers, entity, protocol))
+
+          case MessageStartError(_, info) ⇒
+            throw IllegalResponseException(info)
+        }
+      }
+
+      setHandler(in, idle)
+
+      private def createEntity(creator: EntityCreator[ResponseOutput, ResponseEntity]): ResponseEntity = {
+        creator match {
+          case StrictEntityCreator(entity) ⇒ entity
+          case StreamedEntityCreator(creator) ⇒
+            val entitySource = new SubSourceOutlet[ResponseOutput]("EntitySource")
+
+            entitySource.setHandler(new OutHandler {
+              override def onPull(): Unit = pull(in)
+
+              override def onDownstreamFinish(): Unit = {
+                if (isClosed(in) || eagerClose) {
+                  completeStage()
+                } else {
+                  // drain input to prevent stalling
+                  if (!hasBeenPulled(in)) pull(in)
+                }
+              }
+            })
+
+            setHandler(in, new InHandler {
+              override def onPush(): Unit = grab(in) match {
+                case MessageEnd ⇒
+                  entitySource.complete()
+                  setHandler(in, idle)
+                case messagePart ⇒
+                  entitySource.push(messagePart)
+              }
+
+              override def onUpstreamFinish(): Unit = completeStage()
+            })
+
+            creator(Source.fromGraph(entitySource.source))
+        }
+      }
+
+      override def preStart(): Unit = pull(in)
+      setHandler(out, eagerTerminateOutput)
```

Comment:
https://github.com/akka/akka/commit/ced5aa7ddcac1408e08ed159da9a04fecfad18ed#diff-623ef36775342f2db6b86855f7ffea64R122 uses custom `OutHandler` which just pulls the input, however while testing I've discovered that this leads to port being pulled twice:

```
[ERROR] [01/20/2016 00:58:08.384] [http-streaming-akka.actor.default-dispatcher-4] [ActorSystem(http-streaming)] Outgoing request stream error
java.lang.IllegalArgumentException: requirement failed: Cannot pull port twice
 at scala.Predef$.require(Predef.scala:233)
 at akka.stream.stage.GraphStageLogic.pull(GraphStage.scala:420)
 at akka.http.impl.engine.client.OutgoingConnectionBlueprint$ResponsePrep$$anon$3$$anon$5.onPull(OutgoingConnectionBlueprint.scala:185)
 ...
```

If I understand it correctly this happens because of handler which is set up on `SubSourceOutlet` for entity, which also produces pull's from the same stage.